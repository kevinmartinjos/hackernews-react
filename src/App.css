#app {
    display: grid;
    grid-template: 10% 90%/10% 90%;
}

#Sidebar {
    grid-area: 2/1/2/2;
}

#Navbar {
    grid-area: 1/1/1/2;
}

#MainContent {
    grid-area: 2/2/2/2;
    display: grid;
    grid-template: 10% 80% 10%/ repeat(5, 1fr) 50%;
}

#Search {
    grid-area: 1/1/2/6;
}

#Articles {
    grid-area: 2/1/2/6;
}

#MoreButton {
    grid-area: 3/1/3/1;
}

#CommentBrowser {
    grid-area: 2/5/2/7;
}

#Archives {
    grid-area: 2/1/2/6;
}