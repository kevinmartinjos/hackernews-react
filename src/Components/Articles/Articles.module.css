#Articles {
    grid-area: 2/1/2/6;
}
.spaced {
    margin-right: 5px;
}

.hover:hover {
    color: lightblue;
    cursor: pointer;
}