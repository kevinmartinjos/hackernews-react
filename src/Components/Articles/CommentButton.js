import React, {Component} from 'react';
import {Icon} from 'semantic-ui-react';
import styles from './Articles.module.css';

class CommentButton extends Component {
    constructor(props) {
        super(props);
        this.onPressed = this.onPressed.bind(this);
    }
    onPressed() {
        this.props.onViewComments(this.props.article.objectID);
    }
    render() {
        const {isPressed} = this.props;
        const outline = isPressed ? "": "outline";
        return (
            <Icon name={'comment ' + outline} onClick={this.onPressed} className={styles.clickable}/>
        )
    }
}

export default CommentButton;
