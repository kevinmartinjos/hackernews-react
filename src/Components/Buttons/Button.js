import React from 'react';
import {Button} from 'semantic-ui-react';
import {WithLoading} from "../Loading/Loading";

const MoreButton = ({type="button", onClick, children, className='', size="mini"}) =>
    <Button type={type} onClick={onClick} className={className} id="MoreButton" size={size}>
        {children}
    </Button>;

const ButtonWithLoading = WithLoading(MoreButton);

export {MoreButton, ButtonWithLoading};
export default MoreButton;
