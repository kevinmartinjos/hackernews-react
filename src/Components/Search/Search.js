import React, {Component, createRef} from "react";
import {Form, Button} from 'semantic-ui-react';

class Search extends Component {
    constructor(props) {
        super(props);
        this.searchBoxRef = createRef();
    }
    componentDidMount() {
        if(this.searchBoxRef.current) {
            this.searchBoxRef.current.focus();
        }
    }
    render() {
        const {onSubmit, searchTerm, children, onSearchTermChange} = this.props;
        return (
            <Form onSubmit={onSubmit} id="Search">
                <Form.Group>
                    <Form.Input type="text" value={searchTerm} onChange={onSearchTermChange} ref={this.searchBoxRef} placeholder="Search"/>
                    <Button type="submit"> Go! </Button>
                </Form.Group>
            </Form>
        );
    }
}

export default Search;
