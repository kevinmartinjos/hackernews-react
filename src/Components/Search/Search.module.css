#Search {
    grid-area: 1/1/2/3;
}

.search_box {
    margin: 5px;
    padding: 5px;
}