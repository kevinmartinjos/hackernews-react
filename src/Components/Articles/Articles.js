import React from 'react';
import Button from "../Buttons/Button";

const Articles = ({articles, searchTerm, onDismiss}) =>
    <div className="table">
        {articles.map(article =>
            <div key={article.objectID}>
                <span><a href={article.url}> {article.title} </a></span>
                <span>{article.author} </span>
                <span>{article.num_comments}</span>
                <span>{article.points}</span>
                <span>
                            <Button
                                onClick={() => onDismiss(article.objectID)}
                                className="button-inline"
                            >
                                Dismiss
                            </Button>
                        </span>
            </div>
        )}
    </div>;

export default Articles;
