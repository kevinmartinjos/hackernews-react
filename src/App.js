import React, {Component} from 'react';
import './App.css';
import Search from "./Components/Search/Search";
import Articles from "./Components/Articles/Articles";
import {ButtonWithLoading} from "./Components/Buttons/Button";

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';

class App extends Component {
    constructor(props) {
        super(props);
        this.state = {
            articles: [],
            searchTerm: "",
            page: 0,
            isLoading: false
        };

        this.onDismiss = this.onDismiss.bind(this);
        this.onSearchTermChange = this.onSearchTermChange.bind(this);
        this.setSearchResult = this.setSearchResult.bind(this);
        this.onSearchSubmit = this.onSearchSubmit.bind(this);
        this.fetchSearchResult = this.fetchSearchResult.bind(this);
        this.loadMore = this.loadMore.bind(this);
    }

    onSearchTermChange(event) {
        this.setState({searchTerm: event.target.value});
    }
    onDismiss(objectID) {
        const updatedArticles = this.state.articles.filter(article => article.objectID !== objectID)
        this.setState({articles: updatedArticles})
    }
    setSearchResult(result) {
        const page = result.page;
        let hits = result.hits
        if(page)
            hits = [...this.state.articles, ...hits];
        this.setState({articles: hits, page: result.page, isLoading: false});
    }
    fetchSearchResult(searchTerm, page=0) {
        this.setState({isLoading: true});
        fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`)
            .then(response => response.json())
            .then(result => this.setSearchResult(result))
            .catch(error => error);
    }
    loadMore() {
        const {searchTerm, page} = this.state;
        this.fetchSearchResult(searchTerm, page + 1);
    }
    onSearchSubmit(event) {
        const {searchTerm} = this.state;
        event.preventDefault();
        this.fetchSearchResult(searchTerm);
    }
    componentDidMount() {
        const {searchTerm} = this.state;
        this.fetchSearchResult(searchTerm);
    }
    render() {
        const {searchTerm, articles, isLoading} = this.state;
        return (
            <div className="page">
                <div className="interactions">
                    <Search
                        searchTerm={searchTerm}
                        onSearchTermChange={this.onSearchTermChange}
                        onSubmit={this.onSearchSubmit}
                    >
                        Search title:
                    </Search>
                </div>
                {articles.length
                    ? <Articles articles={articles} searchTerm={searchTerm} onDismiss={this.onDismiss}/>
                    :
                    <div> No results </div>
                }
                <ButtonWithLoading onClick={this.loadMore} isLoading={isLoading}>Moar!</ButtonWithLoading>
            </div>
        );
    }
}

export default App;
