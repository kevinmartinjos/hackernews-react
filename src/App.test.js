import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from "./App";
import Search from "./Components/Search/Search";
import Button from "./Components/Buttons/Button";
import Articles from "./Components/Articles/Articles";


describe('App', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<App/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('has a valid snapshot', () => {
       const component = renderer.create(<App />);
       let tree = component.toJSON();
       expect(tree).toMatchSnapshot();
    });
});


describe('Search', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Search/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('has a valid snapshot', () => {
       const component = renderer.create(<Search/>);
       let tree = component.toJSON();
       expect(tree).toMatchSnapshot();
    });
});


describe('Button', () => {
   it('renders without crashing', () => {
       const div = document.createElement('div');
       ReactDOM.render(<Button> Hello world </Button>, div);
       ReactDOM.unmountComponentAtNode(div);
   }) ;

   test('has a valid snapshot', () => {
       const component = renderer.create(<Button> Hello world </Button>);
       let tree = component.toJSON();
       expect(tree).toMatchSnapshot();
   });
});

describe('Articles', () => {
    const props = {
        articles: [
            {title: 'Hello', author: 'World', num_comments: 2, points: 20, objectID: 1},
            {title: 'Foo', author: 'Bar', num_comments: 6, points: 10, objectID: 2}
        ]
    };

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Articles {...props}/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('has a valid snapshot', () => {
       const component = renderer.create(<Articles {...props}/>);
       let tree = component.toJSON();
       expect(tree).toMatchSnapshot();
    });

});