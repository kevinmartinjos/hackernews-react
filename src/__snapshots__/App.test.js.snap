// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      Search title:
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className=""
        onClick={undefined}
        type="submit"
      >
         Go! 
      </button>
    </form>
  </div>
  <div>
     No results 
  </div>
  <span>
    Loading ... 
  </span>
</div>
`;

exports[`Articles has a valid snapshot 1`] = `
<div
  className="table"
>
  <div>
    <span>
      <a
        href={undefined}
      >
         
        Hello
         
      </a>
    </span>
    <span>
      World
       
    </span>
    <span>
      2
    </span>
    <span>
      20
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div>
    <span>
      <a
        href={undefined}
      >
         
        Foo
         
      </a>
    </span>
    <span>
      Bar
       
    </span>
    <span>
      6
    </span>
    <span>
      10
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={undefined}
  type="button"
>
   Hello world 
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={undefined}
>
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    className=""
    onClick={undefined}
    type="submit"
  >
     Go! 
  </button>
</form>
`;
